src/Data/AIG/CompactGraph.hs

```diff
@@ -0,0 +1,444 @@
+
+------------------------------------------------------------------
+-- AIG parsing
+
+spaces :: Get ()
+spaces = BP.skipWhile BP.isHorizontalSpace
+
+getIntWords :: Get [Int]
+getIntWords = BP.decimal `BP.sepBy` spaces
+
+getIntWordsLine :: Get [Int]
+getIntWordsLine = getIntWords <* BP.skipWhile BP.isEndOfLine
+
+getHeader :: Get (Int, Int, Int, Int, Int)
+getHeader =
+  do BP.string (modeString Binary)
+     spaces
+     ns <- getIntWordsLine
+     case ns of
+       [m, i, l, o, a] -> return (m, i, l, o, a)
+       _ -> fail $ "Invalid AIG header: " ++ show ns
+
+getOutput :: Get (CompactLit s)
+getOutput =
+  do ns <- getIntWordsLine
+     case ns of
+       [n] -> return (CompactLit (fromIntegral n))
+       _ -> fail $ "Invalid output line: " ++ show ns
+
+getDifference :: Get Word32
+getDifference = go 0 0
+  where
+    addByte x b i = x .|. (((fromIntegral b) .&. 0x7f) `shiftL` (7*i))
+    go x i =
+      do b <- getWord8
+         let x' = addByte x b i
+         if b .&. 0x80 /= 0 then go x' (i + 1) else return x'
+
+getDifferences :: Get (Word32, Word32)
+getDifferences = (,) <$> getDifference <*> getDifference
+
+getGraph :: Get (Var, [Var], [CompactLit s], Map Var (CompactLit s, CompactLit s))
+getGraph =
+  do (maxvar, ninputs, nlatches, nouts, nands) <- getHeader
+     outputs <- replicateM (nlatches + nouts) getOutput
+     diffPairs <- replicateM nands getDifferences
+     let maxInput = fromIntegral ninputs
+         inputs = [Var 1 .. Var maxInput]
+         andVars = take (length diffPairs) [(maxInput + 1) ..]
+         addDiff v (ld, rd) = (Var v, (CompactLit l, CompactLit r))
+           where
+             l = (v `shiftL` 1) - ld
+             r = l - rd
+         andAssocs = zipWith addDiff andVars diffPairs
+     return ( Var (fromIntegral maxvar)
+            , reverse inputs
+            , outputs
+            , Map.fromList andAssocs
+            )
+
+------------------------------------------------------------------
+-- Class instances
+
+instance IsLit CompactLit where
+  not (CompactLit x) = CompactLit (x `xor` 1)
+  (===) = (==)
+
+instance IsAIG CompactLit CompactGraph where
+  withNewGraph _proxy k = k =<< newCompactGraph
+
+  aigerNetwork _proxy fp =
+    do (maxv, inps, outs, gates) <- runGet getGraph <$> LBS.readFile fp
+       maxVar  <- newIORef maxv
+       inputs  <- newIORef inps
+       andMap  <- newIORef gates
+       return (Network CompactGraph {..} outs)
+
+  trueLit  _g = CompactLit 1
+  falseLit _g = CompactLit 0
+
+  newInput g =
+    do v <- newVar g
+       modifyIORef' (inputs g) (v :)
+       return (varToLit v)
+
+  and g x y =
+    do v <- newVar g
+       let l = max x y
+           r = min x y
+       modifyIORef' (andMap g) $ Map.insert v (l, r)
+       return (varToLit v)
+
+  inputCount g = length <$> readIORef (inputs g)
+
+  -- | Get input at given index in the graph.
+  getInput _g _i =
+    fail "Function getInput not implemented for CompactGraph"
```

Comment:
Is there a particular reason we can't implement this?

Reply:
Nope! It's just that no clients currently use it. Given that it's in the interface, though, it's probably worth implementing. I'll do that.